torch/csrc/distributed/c10d/OpsImpl.cpp

```diff
@@ -42,6 +68,24 @@ TORCH_LIBRARY_IMPL(c10d, CPU, m) {
 TORCH_LIBRARY_IMPL(c10d, CUDA, m) {
   m.impl("broadcast_", broadcast_cuda_);
 }
+
+TORCH_LIBRARY_IMPL(c10d, CPU, m) {
+  m.impl("allreduce_", allreduce_cpu_);
+}
+
+// TODO: The SparseCPU/SparseCUDA dispatched methods are only used to support
+// sparse all_reduce in the Gloo backend
+TORCH_LIBRARY_IMPL(c10d, SparseCPU, m) {
+  m.impl("allreduce_", allreduce_cpu_);
+}
```

Comment:
nit: is there actually a `SparseCPU` or `SparseCUDA` implementation in Gloo?
If not, do we need to add it here?

Reply:
There is logic within `allreduce()` to handle sparse tensors (https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/ProcessGroupGloo.cpp#L1465-L1467) to which it branches to different logic, so having each implementation just call allreduce_ keeps behavior the same, but I think there may be a cleaner way to do this, we should discuss this.